File: scripts/aggregate.js
#!/usr/bin/env node
/*
 * Deterministic aggregation of per-persona UX verdicts.
 *
 * Usage:  node aggregate.js <verdicts.json>
 *   <verdicts.json> is either an array of persona verdict objects, or
 *   an object shaped { url, mode, personas: [ ...verdicts ] }.
 *
 * Prints a metrics JSON to stdout. All arithmetic (First-Run AX Score,
 * per-dimension averages, friction, retention counts) is computed here so the
 * numbers are exact and reproducible — the LLM is never trusted to do the math.
 */

const fs = require("fs");

const DIMS = [
  "clarity",
  "coldStart",
  "entryBarrier",
  "firstTaskSuccess",
  "ahaReached",
  "nextAction",
];

function fail(msg) {
  console.error(`aggregate.js: ${msg}`);
  process.exit(1);
}

const file = process.argv[2];
if (!file) fail("missing input file argument");

let raw;
try {
  raw = JSON.parse(fs.readFileSync(file, "utf8"));
} catch (e) {
  fail(`could not read/parse ${file}: ${e.message}`);
}

const personas = Array.isArray(raw) ? raw : raw.personas;
if (!Array.isArray(personas) || personas.length === 0) {
  fail("input has no persona verdicts (expected an array or { personas: [...] })");
}

const round1 = (n) => Math.round(n * 10) / 10;
const mean = (arr) => (arr.length ? arr.reduce((a, b) => a + b, 0) / arr.length : null);

// A valid rubric score is an integer-ish number in [1,5].
function validScore(v) {
  return typeof v === "number" && isFinite(v) && v >= 1 && v <= 5;
}

const warnings = [];

// ---- First-Run AX -------------------------------------------------------
const perPersonaAX = [];
const dimValues = Object.fromEntries(DIMS.map((d) => [d, []]));

personas.forEach((p, i) => {
  const id = (p.persona && (p.persona.id || p.persona.name)) || `persona-${i + 1}`;
  const ax = p.firstRunAX || {};
  const dims = {};
  const collected = [];
  for (const d of DIMS) {
    const v = ax[d];
    if (validScore(v)) {
      dims[d] = v;
      dimValues[d].push(v);
      collected.push(v);
    } else {
      warnings.push(`${id}: firstRunAX.${d} missing or out of range (got ${JSON.stringify(v)})`);
    }
  }
  // Recompute the persona score from the dimensions — authoritative, ignores any
  // score the LLM wrote.
  const score = collected.length === DIMS.length ? round1(mean(collected)) : null;
  if (score === null) warnings.push(`${id}: First-Run AX score not computed (incomplete dimensions)`);
  perPersonaAX.push({ id, dims, score });
});

const dimensionAverages = {};
for (const d of DIMS) dimensionAverages[d] = dimValues[d].length ? round1(mean(dimValues[d])) : null;

const personaScores = perPersonaAX.map((p) => p.score).filter((s) => s !== null);
const overallScore = personaScores.length ? round1(mean(personaScores)) : null;

const ranked = Object.entries(dimensionAverages).filter(([, v]) => v !== null).sort((a, b) => a[1] - b[1]);
const weakestDimension = ranked.length ? ranked[0][0] : null;
const strongestDimension = ranked.length ? ranked[ranked.length - 1][0] : null;

// ---- Friction -----------------------------------------------------------
const frictionPer = personas.map((p, i) => {
  const id = (p.persona && (p.persona.id || p.persona.name)) || `persona-${i + 1}`;
  const f = p.overallFriction;
  const valid = typeof f === "number" && isFinite(f) && f >= 0 && f <= 100;
  if (!valid) warnings.push(`${id}: overallFriction missing or out of range (got ${JSON.stringify(f)})`);
  return { id, overallFriction: valid ? f : null };
});
const frictionVals = frictionPer.map((x) => x.overallFriction).filter((x) => x !== null);
const frictionAverage = frictionVals.length ? round1(mean(frictionVals)) : null;

// ---- Retention ----------------------------------------------------------
const retention = { "would-stay": 0, "would-leave": 0, unsure: 0 };
personas.forEach((p) => {
  const v = p.retentionVerdict;
  if (v in retention) retention[v] += 1;
});

const out = {
  personaCount: personas.length,
  firstRunAX: {
    overallScore,
    dimensionAverages,
    weakestDimension,
    strongestDimension,
    perPersona: perPersonaAX.map(({ id, score }) => ({ id, score })),
  },
  friction: { average: frictionAverage, perPersona: frictionPer },
  retention,
  warnings,
};

process.stdout.write(JSON.stringify(out, null, 2) + "\n");

File: scripts/aggregate.py
#!/usr/bin/env python3
"""Deterministic aggregation of per-persona UX verdicts.

Usage:  python aggregate.py <verdicts.json>
  <verdicts.json> is either an array of persona verdict objects, or an object
  shaped { "url", "goal", "personas": [ ...verdicts ] }.

Prints a metrics JSON to stdout. All arithmetic (First-Run AX Score, per-dimension
averages, friction, retention counts) is computed here so the numbers are exact and
reproducible -- the LLM is never trusted to do the math.
"""

import json
import math
import sys

# Force UTF-8 stdout so non-ASCII (e.g. Korean persona ids) survive Windows' cp949 console.
try:
    sys.stdout.reconfigure(encoding="utf-8")
except AttributeError:
    pass

DIMS = [
    "clarity",
    "coldStart",
    "entryBarrier",
    "firstTaskSuccess",
    "ahaReached",
    "nextAction",
]


def fail(msg):
    sys.stderr.write(f"aggregate.py: {msg}\n")
    sys.exit(1)


def round1(n):
    # Round half up (3.25 -> 3.3), not Python's default banker's rounding.
    return math.floor(n * 10 + 0.5) / 10


def mean(arr):
    return sum(arr) / len(arr) if arr else None


def valid_score(v):
    """A valid rubric score is a number in [1, 5]."""
    return isinstance(v, (int, float)) and not isinstance(v, bool) and 1 <= v <= 5


def main():
    if len(sys.argv) < 2:
        fail("missing input file argument")

    path = sys.argv[1]
    try:
        with open(path, "r", encoding="utf-8") as f:
            raw = json.load(f)
    except (OSError, json.JSONDecodeError) as e:
        fail(f"could not read/parse {path}: {e}")

    personas = raw if isinstance(raw, list) else raw.get("personas")
    if not isinstance(personas, list) or len(personas) == 0:
        fail("input has no persona verdicts (expected an array or { personas: [...] })")

    warnings = []

    # ---- First-Run AX ---------------------------------------------------
    per_persona_ax = []
    dim_values = {d: [] for d in DIMS}

    for i, p in enumerate(personas):
        persona = p.get("persona") or {}
        pid = persona.get("id") or persona.get("name") or f"persona-{i + 1}"
        ax = p.get("firstRunAX") or {}
        collected = []
        for d in DIMS:
            v = ax.get(d)
            if valid_score(v):
                dim_values[d].append(v)
                collected.append(v)
            else:
                warnings.append(
                    f"{pid}: firstRunAX.{d} missing or out of range (got {json.dumps(v)})"
                )
        # Recompute the persona score from the dimensions -- authoritative, ignores
        # any score the LLM wrote.
        if len(collected) == len(DIMS):
            score = round1(mean(collected))
        else:
            score = None
            warnings.append(f"{pid}: First-Run AX score not computed (incomplete dimensions)")
        per_persona_ax.append({"id": pid, "score": score})

    dimension_averages = {
        d: (round1(mean(dim_values[d])) if dim_values[d] else None) for d in DIMS
    }

    persona_scores = [p["score"] for p in per_persona_ax if p["score"] is not None]
    overall_score = round1(mean(persona_scores)) if persona_scores else None

    ranked = sorted(
        ((d, v) for d, v in dimension_averages.items() if v is not None),
        key=lambda kv: kv[1],
    )
    weakest_dimension = ranked[0][0] if ranked else None
    strongest_dimension = ranked[-1][0] if ranked else None

    # ---- Friction -------------------------------------------------------
    friction_per = []
    for i, p in enumerate(personas):
        persona = p.get("persona") or {}
        pid = persona.get("id") or persona.get("name") or f"persona-{i + 1}"
        f = p.get("overallFriction")
        is_valid = isinstance(f, (int, float)) and not isinstance(f, bool) and 0 <= f <= 100
        if not is_valid:
            warnings.append(f"{pid}: overallFriction missing or out of range (got {json.dumps(f)})")
        friction_per.append({"id": pid, "overallFriction": f if is_valid else None})

    friction_vals = [x["overallFriction"] for x in friction_per if x["overallFriction"] is not None]
    friction_average = round1(mean(friction_vals)) if friction_vals else None

    # ---- Retention ------------------------------------------------------
    retention = {"would-stay": 0, "would-leave": 0, "unsure": 0}
    for p in personas:
        v = p.get("retentionVerdict")
        if v in retention:
            retention[v] += 1

    out = {
        "personaCount": len(personas),
        "firstRunAX": {
            "overallScore": overall_score,
            "dimensionAverages": dimension_averages,
            "weakestDimension": weakest_dimension,
            "strongestDimension": strongest_dimension,
            "perPersona": per_persona_ax,
        },
        "friction": {"average": friction_average, "perPersona": friction_per},
        "retention": retention,
        "warnings": warnings,
    }

    sys.stdout.write(json.dumps(out, ensure_ascii=False, indent=2) + "\n")


if __name__ == "__main__":
    main()
